core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java

```diff
@@ -1772,6 +1779,98 @@ public Node visitLateral(SqlBaseParser.LateralContext context)
         return new Lateral(getLocation(context), (Query) visit(context.query()));
     }
 
+    @Override
+    public Node visitTableFunctionInvocation(SqlBaseParser.TableFunctionInvocationContext context)
+    {
+        return visit(context.tableFunctionCall());
+    }
+
+    @Override
+    public Node visitTableFunctionCall(SqlBaseParser.TableFunctionCallContext context)
+    {
+        QualifiedName name = getQualifiedName(context.qualifiedName());
+        List<Argument> arguments = visit(context.tableFunctionArgument(), Argument.class);
+        List<List<QualifiedName>> copartitioning = ImmutableList.of();
+        if (context.COPARTITION() != null) {
+            copartitioning = context.copartitionTables().stream()
+                    .map(tablesList -> tablesList.qualifiedName().stream()
+                            .map(this::getQualifiedName)
+                            .collect(toImmutableList()))
+                    .collect(toImmutableList());
+        }
+
+        return new TableFunctionInvocation(getLocation(context), name, arguments, copartitioning);
+    }
+
+    @Override
+    public Node visitTableFunctionArgument(SqlBaseParser.TableFunctionArgumentContext context)
+    {
+        Optional<Identifier> name = visitIfPresent(context.identifier(), Identifier.class);
+        Node value;
+        if (context.tableArgument() != null) {
+            value = visit(context.tableArgument());
+        }
+        else if (context.descriptorArgument() != null) {
+            value = visit(context.descriptorArgument());
+        }
+        else {
+            value = visit(context.expression());
+        }
+
+        return new Argument(getLocation(context), name, value);
+    }
+
+    @Override
+    public Node visitTableArgument(SqlBaseParser.TableArgumentContext context)
+    {
+        Relation table;
+        if (context.qualifiedName() != null) {
+            table = new Table(getLocation(context.TABLE()), getQualifiedName(context.qualifiedName()));
+        }
+        else {
+            table = new TableSubquery(getLocation(context.TABLE()), (Query) visit(context.query()));
```

Comment:
It'd be better to model these as sub-rules so that the construction occurs naturally by the recursive tree traversal instead of switching on whether qualifiedName is present or not